azurelinuxagent/common/osutil/redhat.py

```diff
@@ -140,17 +139,83 @@ def get_nm_controlled(self):
+        if not os.path.isfile(filepath):
+            msg = "Unable to determine primary network interface {0} state, because state file does not exist: {1}".format(ifname, filepath)
+            logger.warn(msg)
+            raise Exception(msg)
+
+        try:
+            nic_operstate = fileutil.read_file(filepath).rstrip().lower()
+            nic_general_state = shellutil.run_command(nic_general_state_cmd, log_error=True).rstrip().lower()
+            if nic_operstate != "up":
+                logger.warn("The primary network interface {0} operational state is '{1}'.".format(ifname, nic_operstate))
+            else:
+                logger.info("The primary network interface {0} operational state is '{1}'.".format(ifname, nic_operstate))
+            if nic_general_state != "100 (connected)":
+                logger.warn("The primary network interface {0} general state is '{1}'.".format(ifname, nic_general_state))
+            else:
+                logger.info("The primary network interface {0} general state is '{1}'.".format(ifname, nic_general_state))
+            return nic_operstate, nic_general_state
+        except Exception as e:
+            msg = "Unexpected error while determining the primary network interface state: {0}".format(e)
+            logger.warn(msg)
+            raise Exception(msg)
+
+    def check_and_recover_nic_state(self, ifname):
+        """
+        Checks if the provided network interface is in an 'up' state. If the network interface is in a 'down' state,
+        attempt to recover the interface by restarting the Network Manager service.
+
+        Raises an exception if an attempt to bring the interface into an 'up' state fails, or if the state
+         of the network interface cannot be determined.
+        """
+        nic_operstate, nic_general_state = self.get_nic_operational_and_general_states(ifname)
+        if nic_operstate == "down" or "disconnected" in nic_general_state:
+            logger.info("Restarting the Network Manager service to recover network interface {0}".format(ifname))
+            self.restart_network_manager()
+            # Interface does not come up immediately after NetworkManager restart. Wait 1 second before checking
+            # network interface state.
+            time.sleep(1)
+            nic_operstate, nic_general_state = self.get_nic_operational_and_general_states(ifname)
+            if nic_operstate != "up" or nic_general_state != "100 (connected)":
+                msg = "Network Manager restart failed to bring network interface {0} into 'up' and 'connected' state".format(ifname)
+                logger.warn(msg)
+                raise Exception(msg)
+            else:
+                logger.info("Network Manager restart successfully brought the network interface {0} into 'up' and 'connected' state".format(ifname))
+        elif nic_operstate != "up" or nic_general_state != "100 (connected)":
```

Comment:
when do we hit this condition? is not nic_operstate != "up" same as "down"?

Reply:
documentation says there is an "unknown" state:

> Interface is in unknown state, neither driver nor userspace has set
> operational state. Interface must be considered for user data as
> setting operational state has not been implemented in every driver.

If driver isn't setting operation state, a restart of the NM is unlikely to fix this, so we are just reporting this state as a warning

Reply:
Added comment for clarification on 187